index.js

```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+const util = require('util');
+const Promise = require('bluebird');
+
+// Custom mysql dialect with deadlock retries
+const Client_MySQL = require('knex/lib/dialects/mysql');
+
+class Client_MySQL_deadlock extends Client_MySQL {
+  _query(connection, obj) {
+    const self = this;
+    const args = arguments;
+    let retryAmount = obj.retryAmount || 3;
+
+    function runQuery() {
+      return Reflect.apply(Client_MySQL.prototype._query, self, args)
+        .catch(error => {
+          if (error.code === 'ER_LOCK_WAIT_TIMEOUT' && --retryAmount > 0) {
+            console.log(`${error.code} returned, retrying, (${retryAmount}) tries remaning: ${util.inspect(args, {depth:2})}`);
+
+            return runQuery();
+          }
+
+          return Promise.reject(error);
```

Comment:
Because we're already in a wrapped thenable, we can just `throw error` and the rejection will be handled for us